File: swift/Sources/SNES/SNESControls.swift
import SwiftUI

public struct SNESControls: View {
    private let engine: SNESEngine
    private let configuration: SNESControllerConfiguration

    public init(engine: SNESEngine, configuration: SNESControllerConfiguration = .init()) {
        self.engine = engine
        self.configuration = configuration
    }

    public var body: some View {
        GeometryReader { proxy in
            let mode = resolvedMode(for: proxy.size)
            let metrics = Metrics(compact: proxy.size.width < 500)
            let palette = Palette(theme: configuration.theme, overrides: configuration.colors)

            Group {
                if mode == .overlay {
                    overlayControls(metrics: metrics, palette: palette)
                } else {
                    gamepadControls(metrics: metrics, palette: palette)
                }
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottom)
        }
        .accessibilityElement(children: .contain)
        .accessibilityLabel("SNES controller")
    }

    private func resolvedMode(for size: CGSize) -> SNESControllerPresentationMode {
        guard configuration.presentationMode == .automatic else { return configuration.presentationMode }
        return size.width > size.height || size.height < 300 ? .overlay : .gamepad
    }

    private func gamepadControls(metrics: Metrics, palette: Palette) -> some View {
        HStack(alignment: .bottom, spacing: 0) {
            dPad(metrics: metrics, palette: palette, opacity: 1)
            Spacer(minLength: metrics.sectionSpacing)
            VStack(spacing: metrics.utilitySpacing) {
                Text(configuration.theme == .superFamicom ? "SUPER FAMICOM" : "SNES")
                    .font(.system(size: 10, weight: .black, design: .rounded))
                    .foregroundStyle(palette.bodyLabel)
                    .frame(maxWidth: .infinity, alignment: .leading)
                utilityButtons(metrics: metrics, palette: palette, opacity: 1)
            }
            Spacer(minLength: metrics.sectionSpacing)
            actionButtons(metrics: metrics, palette: palette, opacity: 1)
        }
        .padding(metrics.bodyPadding)
        .frame(maxWidth: 680)
        .background { controllerBody(palette: palette) }
        .padding(.horizontal, metrics.outerPadding)
        .padding(.bottom, metrics.outerPadding)
    }

    private func overlayControls(metrics: Metrics, palette: Palette) -> some View {
        ZStack(alignment: .bottom) {
            dPad(metrics: metrics, palette: palette, opacity: configuration.overlayOpacity)
                .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottomLeading)
            utilityButtons(metrics: metrics, palette: palette, opacity: configuration.overlayOpacity)
                .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottom)
            actionButtons(metrics: metrics, palette: palette, opacity: configuration.overlayOpacity)
                .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottomTrailing)
        }
        .padding(.horizontal, metrics.outerPadding)
        .padding(.bottom, metrics.outerPadding)
    }

    private func controllerBody(palette: Palette) -> some View {
        let shape = RoundedRectangle(cornerRadius: 24, style: .continuous)
        return ZStack {
            if configuration.theme == .system {
                #if compiler(>=6.2)
                    if #available(iOS 26.0, macOS 26.0, tvOS 26.0, *) {
                        shape.fill(.clear)
                            .glassEffect(.regular.tint(palette.body), in: shape)
                    } else {
                        shape.fill(.ultraThinMaterial)
                            .overlay(shape.fill(palette.body))
                    }
                #else
                    shape.fill(.ultraThinMaterial)
                        .overlay(shape.fill(palette.body))
                #endif
            } else {
                shape.fill(palette.body)
            }
            shape.stroke(.white.opacity(configuration.theme == .system ? 0.28 : 0.12), lineWidth: 1)
        }
        .shadow(color: .black.opacity(0.24), radius: 14, y: 8)
    }

    private func dPad(metrics: Metrics, palette: Palette, opacity: Double) -> some View {
        Grid(horizontalSpacing: 0, verticalSpacing: 0) {
            GridRow {
                Color.clear.frame(width: metrics.direction, height: metrics.direction)
                control(
                    "▲", .up, size: CGSize(width: metrics.direction, height: metrics.direction), shape: .rounded,
                    color: palette.dPad, palette: palette, opacity: opacity)
                Color.clear.frame(width: metrics.direction, height: metrics.direction)
            }
            GridRow {
                control(
                    "◀", .left, size: CGSize(width: metrics.direction, height: metrics.direction), shape: .rounded,
                    color: palette.dPad, palette: palette, opacity: opacity)
                Rectangle().fill(palette.dPad.opacity(opacity)).frame(
                    width: metrics.direction, height: metrics.direction)
                control(
                    "▶", .right, size: CGSize(width: metrics.direction, height: metrics.direction), shape: .rounded,
                    color: palette.dPad, palette: palette, opacity: opacity)
            }
            GridRow {
                Color.clear.frame(width: metrics.direction, height: metrics.direction)
                control(
                    "▼", .down, size: CGSize(width: metrics.direction, height: metrics.direction), shape: .rounded,
                    color: palette.dPad, palette: palette, opacity: opacity)
                Color.clear.frame(width: metrics.direction, height: metrics.direction)
            }
        }
    }

    private func utilityButtons(metrics: Metrics, palette: Palette, opacity: Double) -> some View {
        HStack(spacing: metrics.utilitySpacing) {
            control(
                "SELECT", .select, size: metrics.utility, shape: .capsule, color: palette.utility, palette: palette,
                opacity: opacity)
            control(
                "START", .start, size: metrics.utility, shape: .capsule, color: palette.utility, palette: palette,
                opacity: opacity)
        }
    }

    private func actionButtons(metrics: Metrics, palette: Palette, opacity: Double) -> some View {
        VStack(spacing: metrics.actionSpacing / 2) {
            HStack(spacing: metrics.actionSpacing) {
                control(
                    "L", .l, size: metrics.utility, shape: .capsule, color: palette.utility, palette: palette,
                    opacity: opacity)
                control(
                    "R", .r, size: metrics.utility, shape: .capsule, color: palette.utility, palette: palette,
                    opacity: opacity)
            }
            Grid(horizontalSpacing: metrics.actionSpacing / 2, verticalSpacing: 0) {
                GridRow {
                    Color.clear.frame(width: metrics.action.width, height: metrics.action.height)
                    control(
                        "X", .x, size: metrics.action, shape: .circle, color: palette.action, palette: palette,
                        opacity: opacity)
                    Color.clear.frame(width: metrics.action.width, height: metrics.action.height)
                }
                GridRow {
                    control(
                        "Y", .y, size: metrics.action, shape: .circle, color: palette.action, palette: palette,
                        opacity: opacity)
                    Color.clear.frame(width: metrics.action.width, height: metrics.action.height)
                    control(
                        "A", .a, size: metrics.action, shape: .circle, color: palette.action, palette: palette,
                        opacity: opacity)
                }
                GridRow {
                    Color.clear.frame(width: metrics.action.width, height: metrics.action.height)
                    control(
                        "B", .b, size: metrics.action, shape: .circle, color: palette.action, palette: palette,
                        opacity: opacity)
                    Color.clear.frame(width: metrics.action.width, height: metrics.action.height)
                }
            }
        }
    }

    @ViewBuilder
    private func control(
        _ label: String, _ button: SNESControllerButton, size: CGSize, shape: ControlShape, color: Color,
        palette: Palette, opacity: Double
    ) -> some View {
        #if os(tvOS)
            Button {
                engine.setButton(button, pressed: true)
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.05) { engine.setButton(button, pressed: false) }
            } label: {
                controlLabel(label, size: size, shape: shape, color: color, palette: palette, opacity: opacity)
            }
            .buttonStyle(.plain)
        #else
            controlLabel(label, size: size, shape: shape, color: color, palette: palette, opacity: opacity)
                .gesture(
                    DragGesture(minimumDistance: 0)
                        .onChanged { _ in engine.setButton(button, pressed: true) }
                        .onEnded { _ in engine.setButton(button, pressed: false) }
                )
        #endif
    }

    private func controlLabel(
        _ label: String, size: CGSize, shape: ControlShape, color: Color, palette: Palette, opacity: Double
    ) -> some View {
        Text(label)
            .font(.system(size: label.count == 1 ? 20 : 9, weight: .black, design: .rounded))
            .foregroundStyle(palette.labels)
            .frame(width: size.width, height: size.height)
            .background { controlSurface(shape: shape, color: color, opacity: opacity) }
            .contentShape(shape.swiftUIShape)
            .accessibilityLabel(label)
    }

    @ViewBuilder
    private func controlSurface(shape: ControlShape, color: Color, opacity: Double) -> some View {
        let surface = shape.swiftUIShape
        if configuration.theme == .system {
            #if compiler(>=6.2)
                if #available(iOS 26.0, macOS 26.0, tvOS 26.0, *) {
                    surface.fill(.clear)
                        .glassEffect(.regular.tint(color.opacity(opacity)).interactive(), in: surface)
                } else {
                    materialSurface(surface, color: color, opacity: opacity)
                }
            #else
                materialSurface(surface, color: color, opacity: opacity)
            #endif
        } else {
            surface.fill(color.opacity(opacity))
                .overlay(surface.stroke(.white.opacity(0.1 * opacity), lineWidth: 1))
                .shadow(color: .black.opacity(0.25 * opacity), radius: 3, y: 2)
        }
    }

    private func materialSurface(_ surface: AnyShape, color: Color, opacity: Double) -> some View {
        surface.fill(.ultraThinMaterial)
            .overlay(surface.fill(color.opacity(opacity)))
            .overlay(surface.stroke(.white.opacity(0.32 * opacity), lineWidth: 1))
            .shadow(color: .black.opacity(0.2 * opacity), radius: 5, y: 3)
    }
}

private struct Metrics {
    let direction: CGFloat
    let action: CGSize
    let utility: CGSize
    let sectionSpacing: CGFloat
    let utilitySpacing: CGFloat
    let actionSpacing: CGFloat
    let bodyPadding: CGFloat
    let outerPadding: CGFloat

    init(compact: Bool) {
        direction = compact ? 38 : 48
        action = CGSize(width: compact ? 50 : 64, height: compact ? 50 : 64)
        utility = CGSize(width: compact ? 46 : 58, height: compact ? 26 : 32)
        sectionSpacing = compact ? 4 : 18
        utilitySpacing = compact ? 4 : 10
        actionSpacing = compact ? 8 : 16
        bodyPadding = compact ? 12 : 22
        outerPadding = compact ? 8 : 20
    }
}

private struct Palette {
    let body: Color
    let dPad: Color
    let action: Color
    let utility: Color
    let labels: Color
    let bodyLabel: Color

    init(theme: SNESControllerTheme, overrides: SNESControllerColorOverrides) {
        let defaults: (Color, Color, Color, Color, Color, Color)
        switch theme {
        case .system:
            defaults = (
                .primary.opacity(0.08), .primary.opacity(0.68), .accentColor, .primary.opacity(0.55), .white, .primary
            )
        case .snes:
            defaults = (
                Color(red: 0.70, green: 0.70, blue: 0.68), Color(white: 0.10),
                Color(red: 0.67, green: 0.05, blue: 0.12), Color(white: 0.16), .white, Color(white: 0.15)
            )
        case .superFamicom:
            defaults = (
                Color(white: 0.78), Color(white: 0.16), Color(red: 0.32, green: 0.24, blue: 0.55), Color(white: 0.30),
                .white, Color(white: 0.18)
            )
        }
        body = overrides.body ?? defaults.0
        dPad = overrides.directionalPad ?? defaults.1
        action = overrides.actionButtons ?? defaults.2
        utility = overrides.utilityButtons ?? defaults.3
        labels = overrides.labels ?? defaults.4
        bodyLabel = overrides.bodyLabel ?? defaults.5
    }
}

private enum ControlShape {
    case circle
    case capsule
    case rounded

    var swiftUIShape: AnyShape {
        switch self {
        case .circle: AnyShape(Circle())
        case .capsule: AnyShape(Capsule())
        case .rounded: AnyShape(RoundedRectangle(cornerRadius: 7, style: .continuous))
        }
    }
}

File: android/snes/src/main/kotlin/snes9x/SNESControls.kt
package snes9x

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

@Composable
fun SNESControls(
    engine: SNESEngine,
    modifier: Modifier = Modifier,
    configuration: SNESControllerConfiguration = SNESControllerConfiguration(),
) {
    BoxWithConstraints(modifier = modifier) {
        val mode =
            when (configuration.presentationMode) {
                SNESControllerPresentationMode.Automatic ->
                    if (maxWidth > maxHeight || maxHeight < 300.dp) {
                        SNESControllerPresentationMode.Overlay
                    } else {
                        SNESControllerPresentationMode.Gamepad
                    }
                else -> configuration.presentationMode
            }
        val metrics = ControllerMetrics(compact = maxWidth < 500.dp)
        val palette = controllerPalette(configuration)

        if (mode == SNESControllerPresentationMode.Overlay) {
            OverlayControls(engine, configuration, metrics, palette)
        } else {
            GamepadControls(engine, configuration, metrics, palette)
        }
    }
}

@Composable
private fun GamepadControls(
    engine: SNESEngine,
    configuration: SNESControllerConfiguration,
    metrics: ControllerMetrics,
    palette: ControllerPalette,
) {
    Row(
        modifier =
            Modifier
                .fillMaxSize()
                .padding(horizontal = metrics.outerPadding, vertical = metrics.outerPadding),
        verticalAlignment = Alignment.Bottom,
    ) {
        Row(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .widthIn(max = 680.dp)
                    .shadow(14.dp, RoundedCornerShape(24.dp))
                    .background(palette.body, RoundedCornerShape(24.dp))
                    .border(1.dp, Color.White.copy(alpha = 0.14f), RoundedCornerShape(24.dp))
                    .padding(metrics.bodyPadding),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.Bottom,
        ) {
            DPad(engine, configuration, metrics, palette, 1f)
            Column(
                modifier = Modifier.padding(horizontal = metrics.sectionSpacing),
                verticalArrangement = Arrangement.spacedBy(metrics.utilitySpacing),
            ) {
                Text(
                    text = if (configuration.theme == SNESControllerTheme.SuperFamicom) "SUPER FAMICOM" else "SNES",
                    color = palette.bodyLabel,
                    fontSize = 10.sp,
                    fontWeight = FontWeight.Black,
                )
                UtilityButtons(engine, configuration, metrics, palette, 1f)
            }
            ActionButtons(engine, configuration, metrics, palette, 1f)
        }
    }
}

@Composable
private fun OverlayControls(
    engine: SNESEngine,
    configuration: SNESControllerConfiguration,
    metrics: ControllerMetrics,
    palette: ControllerPalette,
) {
    val opacity = configuration.overlayOpacity.coerceIn(0f, 1f)
    Box(
        modifier =
            Modifier
                .fillMaxSize()
                .padding(horizontal = metrics.outerPadding, vertical = metrics.outerPadding),
    ) {
        DPad(
            engine,
            configuration,
            metrics,
            palette,
            opacity,
            Modifier.align(Alignment.BottomStart),
        )
        UtilityButtons(
            engine,
            configuration,
            metrics,
            palette,
            opacity,
            Modifier.align(Alignment.BottomCenter),
        )
        ActionButtons(
            engine,
            configuration,
            metrics,
            palette,
            opacity,
            Modifier.align(Alignment.BottomEnd),
        )
    }
}

@Composable
private fun DPad(
    engine: SNESEngine,
    configuration: SNESControllerConfiguration,
    metrics: ControllerMetrics,
    palette: ControllerPalette,
    opacity: Float,
    modifier: Modifier = Modifier,
) {
    Column(modifier = modifier, horizontalAlignment = Alignment.CenterHorizontally) {
        ControllerButton(
            "▲",
            SNESButton.Up,
            engine,
            metrics.direction,
            metrics.direction,
            RoundedCornerShape(7.dp),
            palette.directionalPad,
            palette.labels,
            configuration,
            opacity,
        )
        Row(verticalAlignment = Alignment.CenterVertically) {
            ControllerButton(
                "◀",
                SNESButton.Left,
                engine,
                metrics.direction,
                metrics.direction,
                RoundedCornerShape(7.dp),
                palette.directionalPad,
                palette.labels,
                configuration,
                opacity,
            )
            Box(
                Modifier
                    .size(metrics.direction)
                    .background(palette.directionalPad.copy(alpha = palette.directionalPad.alpha * opacity)),
            )
            ControllerButton(
                "▶",
                SNESButton.Right,
                engine,
                metrics.direction,
                metrics.direction,
                RoundedCornerShape(7.dp),
                palette.directionalPad,
                palette.labels,
                configuration,
                opacity,
            )
        }
        ControllerButton(
            "▼",
            SNESButton.Down,
            engine,
            metrics.direction,
            metrics.direction,
            RoundedCornerShape(7.dp),
            palette.directionalPad,
            palette.labels,
            configuration,
            opacity,
        )
    }
}

@Composable
private fun UtilityButtons(
    engine: SNESEngine,
    configuration: SNESControllerConfiguration,
    metrics: ControllerMetrics,
    palette: ControllerPalette,
    opacity: Float,
    modifier: Modifier = Modifier,
) {
    Row(modifier = modifier, horizontalArrangement = Arrangement.spacedBy(metrics.utilitySpacing)) {
        ControllerButton(
            "SELECT",
            SNESButton.Select,
            engine,
            metrics.utilityWidth,
            metrics.utilityHeight,
            RoundedCornerShape(50),
            palette.utilityButtons,
            palette.labels,
            configuration,
            opacity,
        )
        ControllerButton(
            "START",
            SNESButton.Start,
            engine,
            metrics.utilityWidth,
            metrics.utilityHeight,
            RoundedCornerShape(50),
            palette.utilityButtons,
            palette.labels,
            configuration,
            opacity,
        )
    }
}

@Composable
private fun ActionButtons(
    engine: SNESEngine,
    configuration: SNESControllerConfiguration,
    metrics: ControllerMetrics,
    palette: ControllerPalette,
    opacity: Float,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier,
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(metrics.actionSpacing / 2),
    ) {
        Row(horizontalArrangement = Arrangement.spacedBy(metrics.utilitySpacing)) {
            ControllerButton(
                "L",
                SNESButton.L,
                engine,
                metrics.utilityWidth,
                metrics.utilityHeight,
                RoundedCornerShape(50),
                palette.utilityButtons,
                palette.labels,
                configuration,
                opacity,
            )
            ControllerButton(
                "R",
                SNESButton.R,
                engine,
                metrics.utilityWidth,
                metrics.utilityHeight,
                RoundedCornerShape(50),
                palette.utilityButtons,
                palette.labels,
                configuration,
                opacity,
            )
        }
        ControllerButton(
            "X",
            SNESButton.X,
            engine,
            metrics.actionSize,
            metrics.actionSize,
            CircleShape,
            palette.actionButtons,
            palette.labels,
            configuration,
            opacity,
        )
        Row(horizontalArrangement = Arrangement.spacedBy(metrics.actionSize)) {
            ControllerButton(
                "Y",
                SNESButton.Y,
                engine,
                metrics.actionSize,
                metrics.actionSize,
                CircleShape,
                palette.actionButtons,
                palette.labels,
                configuration,
                opacity,
            )
            ControllerButton(
                "A",
                SNESButton.A,
                engine,
                metrics.actionSize,
                metrics.actionSize,
                CircleShape,
                palette.actionButtons,
                palette.labels,
                configuration,
                opacity,
            )
        }
        ControllerButton(
            "B",
            SNESButton.B,
            engine,
            metrics.actionSize,
            metrics.actionSize,
            CircleShape,
            palette.actionButtons,
            palette.labels,
            configuration,
            opacity,
        )
    }
}

@Composable
private fun ControllerButton(
    label: String,
    button: SNESButton,
    engine: SNESEngine,
    width: Dp,
    height: Dp,
    shape: Shape,
    color: Color,
    labelColor: Color,
    configuration: SNESControllerConfiguration,
    opacity: Float,
) {
    val surface = color.copy(alpha = color.alpha * opacity)
    Box(
        modifier =
            Modifier
                .size(width, height)
                .shadow(if (configuration.theme == SNESControllerTheme.System) 5.dp else 3.dp, shape)
                .background(surface, shape)
                .border(1.dp, Color.White.copy(alpha = 0.18f * opacity), shape)
                .semantics { contentDescription = label }
                .pointerInput(button) {
                    detectTapGestures(
                        onPress = {
                            engine.setButton(button, true)
                            tryAwaitRelease()
                            engine.setButton(button, false)
                        },
                    )
                },
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = label,
            color = labelColor,
            fontSize = if (label.length == 1) 20.sp else 9.sp,
            fontWeight = FontWeight.Black,
        )
    }
}

private data class ControllerMetrics(
    val direction: Dp,
    val actionSize: Dp,
    val utilityWidth: Dp,
    val utilityHeight: Dp,
    val sectionSpacing: Dp,
    val utilitySpacing: Dp,
    val actionSpacing: Dp,
    val bodyPadding: Dp,
    val outerPadding: Dp,
) {
    constructor(compact: Boolean) : this(
        direction = if (compact) 38.dp else 48.dp,
        actionSize = if (compact) 50.dp else 64.dp,
        utilityWidth = if (compact) 46.dp else 58.dp,
        utilityHeight = if (compact) 26.dp else 32.dp,
        sectionSpacing = if (compact) 4.dp else 18.dp,
        utilitySpacing = if (compact) 4.dp else 10.dp,
        actionSpacing = if (compact) 8.dp else 16.dp,
        bodyPadding = if (compact) 12.dp else 22.dp,
        outerPadding = if (compact) 8.dp else 20.dp,
    )
}

private data class ControllerPalette(
    val body: Color,
    val directionalPad: Color,
    val actionButtons: Color,
    val utilityButtons: Color,
    val labels: Color,
    val bodyLabel: Color,
)

@Composable
private fun controllerPalette(configuration: SNESControllerConfiguration): ControllerPalette {
    val defaults =
        when (configuration.theme) {
            SNESControllerTheme.System ->
                ControllerPalette(
                    body = MaterialTheme.colorScheme.surfaceContainerHigh.copy(alpha = 0.92f),
                    directionalPad = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.72f),
                    actionButtons = MaterialTheme.colorScheme.primary,
                    utilityButtons = MaterialTheme.colorScheme.secondary,
                    labels = MaterialTheme.colorScheme.onPrimary,
                    bodyLabel = MaterialTheme.colorScheme.onSurface,
                )
            SNESControllerTheme.SNES ->
                ControllerPalette(
                    Color(0xFFB3B3AE),
                    Color(0xFF1A1A1A),
                    Color(0xFFAB0D1F),
                    Color(0xFF292929),
                    Color.White,
                    Color(0xFF262626),
                )
            SNESControllerTheme.SuperFamicom ->
                ControllerPalette(
                    Color(0xFFC7C7C7),
                    Color(0xFF292929),
                    Color(0xFF51408C),
                    Color(0xFF4D4D4D),
                    Color.White,
                    Color(0xFF2E2E2E),
                )
        }
    val overrides = configuration.colors
    return defaults.copy(
        body = overrides.body ?: defaults.body,
        directionalPad = overrides.directionalPad ?: defaults.directionalPad,
        actionButtons = overrides.actionButtons ?: defaults.actionButtons,
        utilityButtons = overrides.utilityButtons ?: defaults.utilityButtons,
        labels = overrides.labels ?: defaults.labels,
        bodyLabel = overrides.bodyLabel ?: defaults.bodyLabel,
    )
}
